### ISSUE TITLE
Consider 2-signoff for merging

### ISSUE THREAD

**Issue by dwijnand** [sbt/zinc#242]
@eed3si9n WDYT?

**Comment by eed3si9n** [comment]
I think it would be ideal, but I'm not sure if we can always get 2 ppls attention
so maybe 2 ppl with a timeout

**Comment by dwijnand** [comment]
How much of a timeout are you thinking?

**Comment by eed3si9n** [comment]
I guess it's up to us. 2 ~ 3 biz days?

**Comment by jvican** [comment]
That sounds good to me. :+1:

**Comment by romanowski** [comment]
Good idea.

**Comment by dwijnand** [comment]
I think 2-3 business days might be pushing it. I don't see why we need to rush changes in. (I was thinking more like 1-2 weeks).

But I see people are happy with it, so let's go with that.

**Comment by dwijnand** [comment]
Done. The 1.0 branch requires pull request reviews before merging.

There's no direct way to encode "require 2-signoff", sadly.

I'm happy if anyone has any suggestions. Thanks.

**Comment by jvican** [comment]
Great, so we have a deal! :smile:
  
To be honest, I think that a contribution cycle of 1-2 weeks would discourage me from contributing to Zinc as an OSS contributor. I think that if something is important in OSS, especifically in a project as important as Zinc, is fast review iterations so that people are motivated to contribute at any moment without being blocked by reviewers. Not only that, but from the perspective of maintainers it would dramatically slow down any work in Zinc. Release early, release often.
  
At least, this has been my experience in OSS projects.

**Comment by dwijnand** [comment]
Don't confuse a 1-2 week review period with a 1-2 week _timeout_ on 2-signoff.

We're talking about the latter not the former.

We all agree to merge things as fast as possible, and release early and often.

**Comment by jvican** [comment]
Rules condition the game, so even if theoretically it's not 1-2 week timeout, it can happen. The very fact that one PR gets delayed 1-2 weeks because of us not having a look at it is terrible -- and should not happen IMO. For it not to happen, it's better to enforce a shorter term (as we have already done) that does not even contemplate the possibility of that event to happen.

**Comment by eed3si9n** [comment]
> We all agree to merge things as fast as possible, and release early and often.

Given relatively strict stability requirement (bincompat for 2 years) and the cost of making mistakes, I'm not sure if I would agree with reviewing as fast or releasing early/often for compilers and build tools. Unless the PR has been discussed in advance, I don't want to set too high expectation for the contributors.

**Comment by jvican** [comment]
I'm actually for releasing early and often, but doing it with care. So, I think that for sbt a good formula would be to release on pull request so that people can try things out immediately. This instant gratification is huge for contributors -- I get drive by it.
  
So, a good schema would be to cut stable releases every X, where X is the number of days decided by maintainers, and then release normal releases that don't ensure stability of any kind in every PR/every night.
  
Lots of important projects use similar release processes, like Mozilla and the Chrome team. I think that for projects having a huge ecosystem, like sbt does, it's really important to release early and often and only promise stability on versions that have been tried out for months.

**Comment by dwijnand** [comment]
> Given relatively strict stability requirement (bincompat for 2 years) and the cost of making mistakes, I'm not sure if I would agree with reviewing as fast or releasing early/often for compilers and build tools.

If you don't agree to reviewing as fast as possible then why did you propose 2-3 business days as a timeout?..

**Comment by eed3si9n** [comment]
> If you don't agree to reviewing as fast as possible then why did you propose 2-3 business days as a timeout?..

Although I don't think we need to process all PRs as fast as possible, I don't want Zinc development to slow down too much given you and I have limited resource we can allocate, for example in the form of OKR. Once the first review completes, I would want the second one to happen relatively soon.
